includes/class-wc-gateway-komoju-ipn-handler.php

```diff
@@ -4,76 +4,93 @@
 
-	protected $notify_url;
+	protected $webhookSecretToken;
 	protected $secret_key;
 	protected $invoice_prefix;
 	/**
 	 * Constructor
 	 */
-	public function __construct( $sandbox = false, $notify_url = '', $secret_key = '', $invoice_prefix = ''  ) {
+	public function __construct( $webhookSecretToken = '', $secret_key = '', $invoice_prefix = ''  ) {
 		add_action( 'woocommerce_api_wc_gateway_komoju', array( $this, 'check_response' ) );
 		add_action( 'valid-komoju-standard-ipn-request', array( $this, 'valid_response' ) );
 
-		$this->sandbox        	= $sandbox;
-		$this->notify_url	  	= $notify_url;
-		$this->secret_key	  	= $secret_key;
-		$this->invoice_prefix	= $invoice_prefix;
+		$this->webhookSecretToken	  	= $webhookSecretToken;
+		$this->secret_key	   	        = $secret_key;
+		$this->invoice_prefix			= $invoice_prefix;
 	}
 
 	/**
 	 * Check for Komoju IPN Response
 	 */
 	public function check_response() {
-		if ( ! empty( $_GET ) && $this->validate_hmac() ) {
-			$posted = wp_unslash( $_GET );
+		$entityBody = file_get_contents('php://input');
 
-			do_action( "valid-komoju-standard-ipn-request", $posted['transaction'] );
+		if ( ! empty( $entityBody ) && $this->validate_hmac($entityBody) ) {
+			$webhookEvent = new WC_Gateway_Komoju_Webhook_Event($entityBody);
+
+			do_action( "valid-komoju-standard-ipn-request", $webhookEvent );
 			exit;
 		}
 		wp_die( "Komoju IPN Request Failure", "Komoju IPN", array( 'response' => 500 ) );
 	}
 
 	/**
 	 * There was a valid response
-	 * @param  array $posted Post data after wp_unslash
+	 * @param  WC_Gateway_Komoju_Webhook_Event $webhookEvent Webhook event data
 	 */
-	public function valid_response( $posted ) {
-		WC_Gateway_Komoju::log( 'External order num: ' . $posted['external_order_num'] );
-		WC_Gateway_Komoju::log( 'Uuid: ' . $posted['uuid'] );
-		WC_Gateway_Komoju::log( 'Payment status: ' . $posted['status'] );
+	public function valid_response( $webhookEvent ) {
+		WC_Gateway_Komoju::log( 'External order num: ' . $webhookEvent->external_order_num() );
+		WC_Gateway_Komoju::log( 'Uuid: ' . $webhookEvent->uuid() );
+		WC_Gateway_Komoju::log( 'Payment status: ' . $webhookEvent->status() );
 
-		$order = $this->get_komoju_order( $posted, $this->invoice_prefix );
+		$order = $this->get_komoju_order( $webhookEvent, $this->invoice_prefix );
 		if ( $order ) {
-			if ( method_exists( $this, 'payment_status_' . $posted['status'] ) ) {
-				call_user_func( array( $this, 'payment_status_' . $posted['status'] ), $order, $posted );
+			switch($webhookEvent->status()) {
+				case "captured":
+					$this->payment_status_captured( $order, $webhookEvent );
+					break;
+				case "authorized":
+					$this->payment_status_authorized( $order, $webhookEvent );
+					break;
+				case "expired":
+					$this->payment_status_expired( $order, $webhookEvent );
+					break;
+				case "cancelled":
+					$this->payment_status_cancelled( $order, $webhookEvent );
+					break;
+				case "failed":

```

Comment:
@thoiberg `payment.failed` can be sent multiple times for one payment. This happens if a card is declined, a user can try again though through the same payment on the KOMOJU hosted page e.g.

1. payment.failed (first time input)
2. payment.failed (second time input)
3. payment.captured

What happens when a payment fails, then is captured later? Will WooCommrece break?

Reply:
@rramsden ah interesting, I didn't realise that was a possible use case. I don't know off the top of my head, I'll check it out and make any changes that may pop up 👍 1+

Reply:
I've removed handling of the `payment.failed` event in ea9bfc7. We discussed this in Slack but to sum up for anyone reading this comment:

Since the `payment.failed` event could be followed by a `payment.captured` (or `payment.authorized` depending on the payment method) we shouldn't be marking the WooCommerce orders as cancelled, since they could still succeed in the future. There's no need at the moment for WooCommerce to be aware of any possible failed payments, so we just won't process the events, and let the eventual state of the payment (success or abandonment/cancellation) be reflected in WooCommerce.